docs/changelog/3043.bugfix.rst

```diff
@@ -0,0 +1 @@
+Ensure that ``get_requires*`` build hooks are called before ``prepare_metadata*``
```

Comment:
Wait, I think we only need to call get_requires_for_build_wheel and not get_requires_for_build_sdist, no?

Reply:
We have to match right? `get_requires_for_build_x` before `prepare_metada_for_build_x` (x either wheel or editable).

Reply:
Yeah, but x is only editable and wheel, not for sdist 🤔

Reply:
Ok, I can rewrite the note to make this more obvious, but I think that the implementation works like that.

Reply:
Well, dunno I need to have a think. Because now you moved setup logic outside of the setup method 🤔 which is not good, but I understand is the easy way to fix it. I'd like to not have setup outside of setup 🤔

Reply:
Yes. It might be worth to re-consider the implementation. Maybe the problem needs to be considered more in a conceptual level.

Please feel free to close this PR if you want to solve the problem in a different way.

(Sorry, pressed the wrong button earlier).